File: hiroo-js/examples/channel_bot.js
// Channel management bot — demonstrates creating / deleting channels in
// categories via the bot API. The bot must have MANAGE_CHANNELS in the
// target guild.
//
// Commands:
//   /mkchannel <category> <name>   creates a text channel inside <category>
//   /mkvoice   <category> <name>   same, but voice
//   /rmchannel <name>              deletes a channel by name
//
//   BOT_TOKEN=... APP_ID=... node examples/channel_bot.js

import { Bot, Intents, defaultIntents, HTTPError } from "../src/index.js";

const BOT_TOKEN = process.env.BOT_TOKEN;
const APP_ID = process.env.APP_ID;
if (!BOT_TOKEN) throw new Error("Set BOT_TOKEN in env");

const bot = new Bot({
  intents: defaultIntents() | Intents.MESSAGE_CONTENT,
  applicationId: APP_ID,
});

async function findCategory(guildId, name) {
  const channels = await bot.http.getGuildChannels(guildId);
  const wanted = name.trim().toLowerCase();
  return channels.find((c) => c.type === "category" && (c.name ?? "").toLowerCase() === wanted);
}

async function findChannel(guildId, name) {
  const channels = await bot.http.getGuildChannels(guildId);
  const wanted = name.trim().toLowerCase().replace(/^#/, "");
  return channels.find((c) => (c.name ?? "").toLowerCase() === wanted);
}

bot.on("ready", () => {
  console.log(`Logged in as ${bot.user?.username} (id=${bot.user?.id})`);
});

bot.slashCommand(
  { name: "mkchannel", description: "Create a text channel inside a category" },
  async (ctx, { category, name }) => {
    if (!ctx.guildId) return ctx.respond("Команда работает только в сервере.");
    const cat = await findCategory(ctx.guildId, category);
    if (!cat) return ctx.respond(`Категория «${category}» не найдена.`);
    try {
      const ch = await bot.http.createChannel(ctx.guildId, name, { type: "text", parentId: cat.id });
      await ctx.respond(`Канал **#${ch.name}** создан в категории «${cat.name}».`);
    } catch (e) {
      if (e instanceof HTTPError) await ctx.respond(`Не удалось создать канал: ${e.message}`);
      else throw e;
    }
  },
  [
    { name: "category", type: "string", required: true },
    { name: "name", type: "string", required: true },
  ],
);

bot.slashCommand(
  { name: "mkvoice", description: "Create a voice channel inside a category" },
  async (ctx, { category, name }) => {
    if (!ctx.guildId) return ctx.respond("Команда работает только в сервере.");
    const cat = await findCategory(ctx.guildId, category);
    if (!cat) return ctx.respond(`Категория «${category}» не найдена.`);
    try {
      const ch = await bot.http.createChannel(ctx.guildId, name, { type: "voice", parentId: cat.id });
      await ctx.respond(`Голосовой канал **${ch.name}** создан в категории «${cat.name}».`);
    } catch (e) {
      if (e instanceof HTTPError) await ctx.respond(`Не удалось создать канал: ${e.message}`);
      else throw e;
    }
  },
  [
    { name: "category", type: "string", required: true },
    { name: "name", type: "string", required: true },
  ],
);

bot.slashCommand(
  { name: "rmchannel", description: "Delete a channel by name" },
  async (ctx, { name }) => {
    if (!ctx.guildId) return ctx.respond("Команда работает только в сервере.");
    const ch = await findChannel(ctx.guildId, name);
    if (!ch) return ctx.respond(`Канал «${name}» не найден.`);
    try {
      await bot.http.deleteChannel(ctx.guildId, ch.id);
      await ctx.respond(`Канал **#${ch.name}** удалён.`);
    } catch (e) {
      if (e instanceof HTTPError) await ctx.respond(`Не удалось удалить канал: ${e.message}`);
      else throw e;
    }
  },
  [{ name: "name", type: "string", required: true }],
);

bot.run("Bot " + BOT_TOKEN).catch((e) => {
  console.error(e);
  process.exit(1);
});

File: hiroo-py/examples/channel_bot.py
"""Demonstrates channel management via the bot API.

Commands:
  /mkchannel <category-name> <channel-name>   creates a text channel inside
                                              the named category.
  /mkvoice <category-name> <channel-name>     same but for a voice channel.
  /rmchannel <channel-name>                   deletes a channel by name.

The bot must have the MANAGE_CHANNELS permission in the server.

Invoke as:
    BOT_TOKEN=... APP_ID=... python channel_bot.py
"""
import os
import asyncio
import hiroo


BOT_TOKEN = os.environ["BOT_TOKEN"]
APP_ID = os.environ.get("APP_ID")

bot = hiroo.Bot(
    intents=hiroo.Intents.default() | hiroo.Intents.MESSAGE_CONTENT,
    application_id=APP_ID,
)


async def find_category(guild_id: str, name: str) -> dict | None:
    """Locate a category channel in the given guild by (case-insensitive) name."""
    channels = await bot.http.get_guild_channels(guild_id)
    wanted = name.strip().lower()
    for ch in channels:
        if ch.get("type") == "category" and (ch.get("name") or "").lower() == wanted:
            return ch
    return None


async def find_channel(guild_id: str, name: str) -> dict | None:
    channels = await bot.http.get_guild_channels(guild_id)
    wanted = name.strip().lower().lstrip("#")
    for ch in channels:
        if (ch.get("name") or "").lower() == wanted:
            return ch
    return None


@bot.event
async def on_ready():
    print(f"Logged in as {bot.user.username} (id={bot.user.id})")


@bot.slash_command(description="Create a text channel inside a category")
async def mkchannel(ctx: hiroo.CommandContext, category: str, name: str):
    guild_id = (ctx.interaction or {}).get("guild_id") or (ctx.message and ctx.message._raw.get("server_id"))
    if not guild_id:
        await ctx.respond("Работает только в серверном канале.")
        return
    cat = await find_category(guild_id, category)
    if not cat:
        await ctx.respond(f"Категория «{category}» не найдена.")
        return
    try:
        ch = await bot.http.create_channel(
            guild_id, name, type="text", parent_id=cat["id"],
        )
    except hiroo.HTTPError as e:
        await ctx.respond(f"Не удалось создать канал: {e}")
        return
    await ctx.respond(f"Канал **#{ch['name']}** создан в категории «{cat['name']}».")


@bot.slash_command(description="Create a voice channel inside a category")
async def mkvoice(ctx: hiroo.CommandContext, category: str, name: str):
    guild_id = (ctx.interaction or {}).get("guild_id") or (ctx.message and ctx.message._raw.get("server_id"))
    if not guild_id:
        await ctx.respond("Работает только в серверном канале.")
        return
    cat = await find_category(guild_id, category)
    if not cat:
        await ctx.respond(f"Категория «{category}» не найдена.")
        return
    try:
        ch = await bot.http.create_channel(
            guild_id, name, type="voice", parent_id=cat["id"],
        )
    except hiroo.HTTPError as e:
        await ctx.respond(f"Не удалось создать канал: {e}")
        return
    await ctx.respond(f"Голосовой канал **{ch['name']}** создан в категории «{cat['name']}».")


@bot.slash_command(description="Delete a channel by name")
async def rmchannel(ctx: hiroo.CommandContext, name: str):
    guild_id = (ctx.interaction or {}).get("guild_id") or (ctx.message and ctx.message._raw.get("server_id"))
    if not guild_id:
        await ctx.respond("Работает только в серверном канале.")
        return
    ch = await find_channel(guild_id, name)
    if not ch:
        await ctx.respond(f"Канал «{name}» не найден.")
        return
    try:
        await bot.http.delete_channel(guild_id, ch["id"])
    except hiroo.HTTPError as e:
        await ctx.respond(f"Не удалось удалить канал: {e}")
        return
    await ctx.respond(f"Канал **#{ch['name']}** удалён.")


if __name__ == "__main__":
    bot.run("Bot " + BOT_TOKEN)
